lib/AnalysisStructured/PtrAnalysis.cpp

```diff
@@ -0,0 +1,1128 @@
+  // this condition is to handle cases where tt.broadcast and tt.splat are
+  // folded
+  auto attr = cast<DenseElementsAttr>(op.getValue());
+  auto elementType = attr.getElementType();
+  assert(attr.isSplat() && elementType.isa<IntegerType>());
+  auto values = attr.getValues<IntegerAttr>();
+  auto value = values[0].getValue();
+  auto constAttr = builder.getIndexAttr(value.getSExtValue());
+  auto constOp = arith::ConstantOp::materialize(builder, constAttr,
+                                                builder.getIndexType(), loc);
+
+  state.scalar = constOp;
+
+  auto resultType = cast<ShapedType>(op.getResult().getType());
+  for (size_t i = 0; i < resultType.getShape().size(); i++) {
+    if (i == 0) {
+      state.offsets.push_back(constOp.getResult());
+    } else {
+      state.offsets.push_back(builder.getIndexAttr(0));
+    }
+
+    state.sizes.push_back(builder.getIndexAttr(resultType.getShape()[i]));
+    state.strides.push_back(builder.getIndexAttr(0));
+    state.modulos.push_back(builder.getIndexAttr(0));
+  }
+
+  return success();
+}
+
+LogicalResult PtrAnalysis::visitOperandMakeTPtr(tts::MakeTensorPtrOp makeTPtrOp,
+                                                PtrSState &state,
+                                                const Location loc,
+                                                OpBuilder &builder) {
+
+  assert(state.isEmpty());
+  state.source = makeTPtrOp.getBase();
+  state.offsets = makeTPtrOp.getMixedOffsets();
+  state.sizes = makeTPtrOp.getMixedSizes();
+  state.strides = makeTPtrOp.getMixedStrides();
+  state.modulos = makeTPtrOp.getMixedParentSizes();
+
+  return success();
+}
+
+LogicalResult PtrAnalysis::visitOperand(Value operand, PtrSState &state,
+                                        const Location loc,
+                                        OpBuilder &builder) {
+
+  if (knownPtrs.find(operand) != knownPtrs.end()) {
+    state = knownPtrs.lookup(operand);
+    return success();
+  }
+
+  if (operand.getType().isa<IntegerType>()) {
+    OpBuilder::InsertionGuard guard(builder);
+    if (!operand.isa<BlockArgument>() && operand.getDefiningOp()) {
+      builder.setInsertionPointAfter(operand.getDefiningOp());
+    }
+    auto castOp = builder.create<arith::IndexCastOp>(
+        loc, builder.getIndexType(), operand);
+    state.scalar = castOp.getResult();
+    return success();
+  } else if (operand.getType().isa<IndexType>()) {
+    state.scalar = operand;
+    return success();
+  }
+
+  if (operand.getType().isa<triton::PointerType>()) {
+    // A scalar pointer can either be produced by AddPtrOp or a block
+    // argument
+    if (auto op = operand.getDefiningOp()) {
+      if (auto addPtrOp = dyn_cast<triton::AddPtrOp>(op)) {
+        return visitOperandAddptr(cast<triton::AddPtrOp>(op), state, loc,
+                                  builder);
+      } else if (auto makeTensorOp = dyn_cast<triton::MakeTensorPtrOp>(op)) {
+        llvm_unreachable("NYI");
+      } else {
```

Comment:
let's add a github issue to track this

Reply:
This is actually no longer needed. Will change the error message.